Replace threshold with topOffset and bottomOffset

jmeas opened the pull request:
This is a WIP to resolve https://github.com/brigade/react-waypoint/issues/77

It has two notable changes:

1. `threshold` has been split into `offsetTop` and `offsetBottom`
2. The new options accept strings and numbers. Strings can be of the form `"100"`, `"100px"` or `"20%"`. Actual numbers, like `20`, are assumed to be pixel values

---

I haven't added tests 'cause I figure there might be feedback on the implementation. Once the implementation is :+1:'d I'll update the tests.

jmeas commented on src/waypoint.jsx:
```diff
@@ -32,6 +45,20 @@ function debugLog() {
   console.log(arguments); // eslint-disable-line no-console
 }
 
+function parseOffsetAsPixels(str) {
```
lmk if you'd rather these be instance props on the components @trotzig

trotzig replied:
No, this is better. :)

trotzig replied:
On second thought, maybe they should be. I don't think computing these over and over affect performance too much, but it's an easy change.

jmeas replied:
I'll move em in and we can see how it looks 🙂

jmeas replied:
> I don't think computing these over and over affect performance too much, but it's an easy change.

As an aside, I don't think either approach causes anything to be computed or created more than the other.

Most JS module loaders cache the modules, so these will be created the first time that the lib is required and never again. And if we move it into the class, they'll be defined on the prototype, also just once. So it should just be a matter of preference, which could be an aesthetic preference or one based on whether it makes sense to expose these to consumers of the lib.

trotzig replied:
Oh, I thought you were talking about memoizing the value. I was thinking we could store the computed value in an instance variable and only call the method(s) once. But I realize now that we can't do that because of the `contextHeight` being variable. Sorry for the confusion.

jmeas replied:
Ah, okay! I misunderstood - sorry 🙂

jmeas commented on src/waypoint.jsx:
```diff
@@ -202,23 +253,24 @@ export default class Waypoint extends React.Component {
 
-    if (contextScrollTop <= waypointTop + thresholdPx &&
-        waypointTop - thresholdPx <= contextBottom) {
+    if (contextScrollTop <= waypointTop - topOffsetPx &&
```
I'll likely need to adjust the relational operators here. Getting unexpected results when just flipping the `+` to `-`.

trotzig commented on src/waypoint.jsx:
```diff
@@ -18,10 +18,23 @@ const propTypes = {
   scrollableAncestor: PropTypes.any,
+  // `topOffset` can either be a number, in which case its a distance from the
+  // top of the container, or a percentage, in which case it's a percentage of
+  // the height of the visible part of the scrollable ancestor (e.g. 0.1)
```
The `e.g. 0.1` part is confusing.

jmeas replied:
I'll update the comment.

trotzig commented:
This looks good! You may go ahead and fix tests. 👍

jmeas commented:
> This looks good! You may go ahead and fix tests

It might take a few days, but it's on my todo list!

jmeas commented:
It's been some time, but I've rebased this off master and made the existing tests pass (with a small tweak to the current test for threshold). I still need to write the rest of the offset tests (offset top + pixel values for offset), but I figured I'd post this update :v:

trotzig commented:
Cool, this is coming together nicely.

jmeas commented:
@trotzig I updated the JSDoc comments and added tests for all of the different ways to specify the bottom offset. I'd love to get any feedback you have on the JSDoc comments.

Would you also like to see analogous tests for the top offset prop?

trotzig commented:
Sorry for the delay. Let me have a look at this tomorrow when my eyes are less tired. 🌃

jmeas commented:
Sure thing. Thanks!

trotzig commented on src/waypoint.jsx:
```diff
@@ -184,6 +198,62 @@ export default class Waypoint extends React.Component {
 
   /**
+   * @param {string|number} offset Blah blah blah
+   * @param {number} contextHeight Blah blah blah
```
lol, you might want to change this description a little. I think just removing the blah-blah is fine too.

jmeas replied:
Hah. Oops. I missed this blah blah. I'll update.

trotzig commented on src/waypoint.jsx:
```diff
@@ -206,23 +276,26 @@ export default class Waypoint extends React.Component {
     }
-    const thresholdPx = contextHeight * this.props.threshold;
+
+    const {bottomOffset, topOffset} = this.props;
```
No big deal, but we've mostly used spaces inside curlies: 

```js
const { bottomOffset, topOffset } = ...
```

jmeas replied:
Consistency is good – I'll fix this up real quick.

trotzig commented:
All good. If it's not a big deal, we should probably have a simple test for topOffset as well. Don't flesh it out like you did for bottomOffset though. We're probably fine with a single test that checks for onEnter being called.

jmeas commented:
I'll put that together tomorrow.

jmeas commented:
Woops – took a few days longer than expected, but this has been updated. The "blah blah"s are gone, and there's a new test for `topOffset`. I also squashed the commits.

@trotzing, the current setup of the block of the tests that the `bottomOffset` tests made it easier to test for `onLeave` than `onEnter`, so I went ahead and did that.

Lmk if there's anything else I should do/change here :)

trotzig commented:
Thanks! I'm releasing this as 3.0.0 (going from 2.0.3, since this is a breaking change). Stay tuned.

jmeas commented:
Rad. Thanks 🙂
